File: 2024 April/Gray to Binary equivalent/main.py
# Link: https://www.geeksforgeeks.org/problems/gray-to-binary-equivalent-1587115620/1


class Solution:
    def grayToBinary(self, n):
        def toBinary(gray):
            binary = gray[0]
            for i in range(1, len(gray)):
                binary += str(int(binary[-1]) ^ int(gray[i]))
            return binary

        def toBinaryFromDec(n):
            return bin(n)[2:]

        def toDecimal(binary):
            binary = binary[::-1]
            res = 0
            for i in range(len(binary)):
                res += int(binary[i]) * (2**i)
            return res

        def toGray(decimal):
            binary = toBinaryFromDec(decimal)
            gray = binary[0]
            for i in range(1, len(binary)):
                gray += str(int(binary[i]) ^ int(binary[i - 1]))
            return gray

        return toDecimal(toBinary(toBinary(toGray(n))))


if __name__ == "__main__":
    obj = Solution()
    n = 4
    print(obj.grayToBinary(n))
    n = 15
    print(obj.grayToBinary(n))

File: 2024 April/Gray to Binary equivalent/main.cpp
// Link: https://www.geeksforgeeks.org/problems/gray-to-binary-equivalent-1587115620/1

#include <bits/stdc++.h>
using namespace std;

class Solution
{
public:
    int grayToBinary(int n)
    {
        vector<int> ans(32, 0);
        int i = 0;
        while (n > 0)
        {
            ans[i] = n % 2;
            i++;
            n = n / 2;
        }
        reverse(ans.begin(), ans.end());

        vector<int> res(32, 0);

        int sum = ans[0];
        for (int i = 0; i < ans.size(); i++)
        {
            sum = sum ^ ans[i];
            res[i] = sum;
        }
        sum = 0;
        int cnt = 0, j = 0;
        while (j < res.size())
        {
            sum = sum * 2 + res[j];
            j++;
        }
        return sum;
    }
};

int main()
{
    Solution obj;
    int n = 4;
    cout << obj.grayToBinary(n) << endl;
    n = 15;
    cout << obj.grayToBinary(n) << endl;
    return 0;
}